scripts/notify-suppliers-whether-application-made-for-framework.py

```diff
@@ -41,31 +55,48 @@
     arguments = docopt(__doc__)
 
     STAGE = arguments['<stage>']
-    FRAMEWORK_SLUG = arguments['<framework_slug>']
-    GOVUK_NOTIFY_API_KEY = arguments['<govuk_notify_api_key>']
+    FRAMEWORK_SLUG = arguments['<framework>']
+    GOVUK_NOTIFY_API_KEY = arguments['<notify_api_key>']
     DRY_RUN = arguments['--dry-run']
 
+    if arguments["--supplier-ids-from"]:
+        supplier_ids = get_supplier_ids_from_file(arguments["--supplier-ids-from"])
+    elif arguments["--supplier-id"]:
+        try:
+            supplier_ids = tuple(map(int, arguments["--supplier-id"]))
+        except ValueError:
+            raise TypeError("argument to --supplier-id should be an integer")
+    else:
+        supplier_ids = None
+
     mail_client = scripts_notify_client(GOVUK_NOTIFY_API_KEY, logger=logger)
     api_client = DataAPIClient(base_url=get_api_endpoint_from_stage(STAGE),
                                auth_token=get_auth_token('api', STAGE))
 
     framework_data = api_client.get_framework(FRAMEWORK_SLUG)['frameworks']
     update_framework_with_formatted_dates(framework_data)
 
-    context_helper = AppliedToFrameworkSupplierContextForNotify(api_client,
-                                                                FRAMEWORK_SLUG,
-                                                                framework_data['intentionToAwardAt'])
+    context_helper = AppliedToFrameworkSupplierContextForNotify(
+        api_client, FRAMEWORK_SLUG, framework_data['intentionToAwardAt'], supplier_ids
+    )
     context_helper.populate_data()
-    context_data = context_helper.get_users_personalisations()
+    prefix = "[Dry Run] " if DRY_RUN else ""
     error_count = 0
-    for user_email, personalisation in context_data.items():
-        template_key = 'application_made' if personalisation['applied'] else 'application_not_made'
+    for supplier_id, users in context_helper.get_suppliers_with_users_personalisations():
+        logger.info(f"{prefix}Supplier '{supplier_id}'")
+
+        for user, personalisation in users:
+            user_email = user["email address"]
+            template_key = 'application_made' if personalisation['applied'] else 'application_not_made'
+            template = NOTIFY_TEMPLATES[template_key]
+
+            logger.info("{}Sending '{}' email to {}".format(prefix, template_key, hash_string(user_email)))
+
+            if DRY_RUN:
+                continue
 
-        prefix = "[Dry Run] " if DRY_RUN else ""
-        logger.info("{}Sending '{}' email to {}".format(prefix, template_key, hash_string(user_email)))
-        if not DRY_RUN:
             try:
-                mail_client.send_email(user_email, NOTIFY_TEMPLATES[template_key], personalisation, allow_resend=False)
+                mail_client.send_email(user_email, template, personalisation, allow_resend=False)
             except EmailError as e:
                 logger.error(u'Error sending email to {}: {}'.format(hash_string(user_email), e))
```

Comment:
Can we include the supplier ID in the error log, so if it fails we know which ID to re-run with?

Reply:
Good idea